src/app/clusters/commissioner-control-server/commissioner-control-server.cpp

```diff
@@ -0,0 +1,277 @@
+        {
+            ChipLogError(Zcl, "ReverseCommissionNode error: %s", err.AsString());
+        }
+    }
+    else
+    {
+        ChipLogError(Zcl, "No delegate available for ReverseCommissionNode");
+    }
+
+    delete info;
+}
+
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace CommissionerControl {
+
+CommissionerControlServer CommissionerControlServer::mInstance;
+
+CommissionerControlServer & CommissionerControlServer::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR CommissionerControlServer::Init(Delegate & delegate)
+{
+    mDelegate = &delegate;
+    return CHIP_NO_ERROR;
+}
+
+Protocols::InteractionModel::Status CommissionerControlServer::GetSupportedDeviceCategoriesValue(
+    EndpointId endpoint, BitMask<SupportedDeviceCategoryBitmap> * supportedDeviceCategories) const
+{
+    Status status = Attributes::SupportedDeviceCategories::Get(endpoint, supportedDeviceCategories);
+    if (status != Status::Success)
+    {
+        ChipLogProgress(Zcl, "CommissionerControl: reading  supportedDeviceCategories, err:0x%x", to_underlying(status));
+    }
+    return status;
+}
+
+Protocols::InteractionModel::Status
+CommissionerControlServer::SetSupportedDeviceCategoriesValue(EndpointId endpoint,
+                                                             const BitMask<SupportedDeviceCategoryBitmap> supportedDeviceCategories)
+{
+    Status status = Status::Success;
+
+    if ((status = Attributes::SupportedDeviceCategories::Set(endpoint, supportedDeviceCategories)) != Status::Success)
+    {
+        ChipLogProgress(Zcl, "CommissionerControl: writing supportedDeviceCategories, err:0x%x", to_underlying(status));
+        return status;
+    }
+
+    return status;
+}
+
+CHIP_ERROR
+CommissionerControlServer::GenerateCommissioningRequestResultEvent(const Events::CommissioningRequestResult::Type & result)
+{
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(result, kRootEndpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "CommissionerControl: Unable to emit CommissioningRequestResult event: %s", error.AsString());
+    }
+
+    return error;
+}
+
+} // namespace CommissionerControl
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+bool emberAfCommissionerControlClusterRequestCommissioningApprovalCallback(
+    app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
+    const Clusters::CommissionerControl::Commands::RequestCommissioningApproval::DecodableType & commandData)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Status status  = Status::Success;
+
+    ChipLogProgress(Zcl, "Received command to request commissioning approval");
+
+    auto sourceNodeId = getNodeId(commandObj);
+
+    // Check if the command is executed via a CASE session
+    if (sourceNodeId == kUndefinedNodeId)
+    {
+        ChipLogError(Zcl, "Command not executed via CASE session, failing with UNSUPPORTED_ACCESS");
```

Comment:
We already logged an error for this in getNodeId.  Why are we logging both places?

Also, I strongly suggest saying _which_ command.